internal/util-logging/src/main/scala/sbt/internal/util/ConsoleAppender.scala

```diff
@@ -356,18 +357,38 @@ object ConsoleAppender {
  *
  * This logger is not thread-safe.
  */
-class ConsoleAppender private[ConsoleAppender] (
-    name: String,
-    properties: Properties,
-    suppressedMessage: SuppressedTraceContext => Option[String]
-) extends AbstractAppender(name, null, LogExchange.dummyLayout, true, Array.empty) {
-  def this(
-      name: String,
-      out: ConsoleOut,
-      ansiCodesSupported: Boolean,
-      useFormat: Boolean,
-      suppressedMessage: SuppressedTraceContext => Option[String]
-  ) = this(name, Properties.from(out, ansiCodesSupported, useFormat), suppressedMessage)
+class ConsoleAppenderImpl(
+    override private[sbt] val name: String,
+    override private[sbt] val properties: Properties,
+    override private[sbt] val suppressedMessage: SuppressedTraceContext => Option[String]
+) extends ConsoleAppender {
+  private[this] val log4j = new AtomicReference[XAppender](null)
+  override private[sbt] lazy val toLog4J = log4j.get match {
+    case null =>
+      log4j.synchronized {
+        log4j.get match {
+          case null =>
+            val l = new Log4JConsoleAppender(name, properties, suppressedMessage, { event =>
+              val level = ConsoleAppender.toLevel(event.getLevel)
+              val message = event.getMessage
+              try appendMessage(level, message)
+              catch { case _: ClosedChannelException => }
+            })
+            log4j.set(l)
+            l
+          case l => l
+        }
+      }
+  }
+  override def close(): Unit = log4j.get match {
+    case null =>
+    case a    => a.stop()
+  }
+}
+trait ConsoleAppender extends AutoCloseable {
```

Comment:
Would it make sense to call this `Appender`? `ConsoleAppender` sounds like it's a concrete appender targeting the console.

Reply:
Yeah. That was how I first did it but I reverted to `ConsoleAppender` in an attempt to minimize the diff. I'd be happy to switch back to `Appender` which is a better name for the reason you described.

Reply:
Yea. I'd suggest `Appender` to be an interface trait with no implementation and `ConsoleAppenderImpl` could become `ConsoleAppender`?

Reply:
Done https://github.com/sbt/sbt/compare/2dc159fe8c7bb68f473a84ec8ccd7f53feefb71a..a8c1b3160badb470ceaaacfeec7bea46082a5703.